### PR TITLE
Make unified collection serialization opt-in instead of opt-out

### PR DESCRIPTION
**Pull request by szeiger**
Fixes https://github.com/scala/bug/issues/11192

### CODE REVIEW

**Comment by szeiger** [comment]
Any takers? This looks big but it's really just factoring out `Iterable.writeReplace()` (and its overrides for the other standard collection kinds) into a `DefaultSerializable` trait.

**Comment by adriaanm** [comment]
@SethTisue, could you look into this one?

File: src/library/scala/collection/generic/DefaultSerializationProxy.scala [hunk 1]
```diff
@@ -67,3 +67,21 @@ final class DefaultSerializationProxy[A](factory: Factory[A, Any], @transient pr
...
+  protected[this] def writeReplace(): AnyRef = {
+    val f: Factory[Any, Any] = this match {
+      case it: scala.collection.SortedMap[_, _] => it.sortedMapFactory.sortedMapFactory[Any, Any](it.ordering.asInstanceOf[Ordering[Any]]).asInstanceOf[Factory[Any, Any]]
+      case it: scala.collection.Map[_, _] => it.mapFactory.mapFactory[Any, Any].asInstanceOf[Factory[Any, Any]]
+      case it: scala.collection.SortedSet[_] => it.sortedIterableFactory.evidenceIterableFactory[Any](it.ordering.asInstanceOf[Ordering[Any]])
+      case it => it.iterableFactory.iterableFactory
+    }
+    new DefaultSerializationProxy(f, this)
+  }
+}
```
Comment (julienrf) [inline src/library/scala/collection/generic/DefaultSerializationProxy.scala:87]: If I understand correctly, developers who want to implement a custom collection type can mix this trait to support the proxy-based serialization?

~~~ scala
class MyCollection[A]
  extends Iterable[A]
    with IterableOps[A, MyCollection, MyCollection[A]]
    with DefaultSerializable { … }
~~~

Comment (szeiger) [reply src/library/scala/collection/generic/DefaultSerializationProxy.scala:87]: Yes, that's what most of the collection classes are doing now (except from some wrapper classes like the Java/Scala converters which use the default serialization)

**Review by SethTisue — APPROVED** [review]
needs (trivial) rebase. after that, let's merge this.

**Comment by adriaanm** [comment]
rebased

**Pull request merged**